core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -429,10 +434,19 @@ public FetchSearchResult executeFetchPhase(ShardFetchRequest request, SearchTask
             throw ExceptionsHelper.convertToRuntime(e);
         } finally {
             cleanContext(context);
+            //Release searchContext if we have no more hits
+            if ((context.queryResult().size()) >= context.scrollContext().totalHits) {
+                if (context.scrollContext() != null) {
+                    releasedContexts.add(context.id());
+                    freeContext(context.id());
```

Comment:
can you add a comment about why you add to `releasedContexts` first? I think it is important for correctness?